Dockerfile

```diff
@@ -7,25 +12,87 @@ RUN apt-get update && apt-get install -y wget \
     git \
     python
 
-ADD . /opt/apalache/
+ADD ./3rdparty /opt/apalache/3rdparty
 WORKDIR /opt/apalache/
 
 # Workaround for Surefire not finding ForkedBooter
 # (see https://stackoverflow.com/questions/53010200/maven-surefire-could-not-find-forkedbooter-class)
 ENV _JAVA_OPTIONS="-Djdk.net.URLClassPath.disableClassPathURLCheck=true"
 
 RUN 3rdparty/install-local.sh --nocache
+
+FROM maven:3.6.3-jdk-8-slim AS builder
+
+# This is the image for building apalache itself
+
+ADD . /opt/apalache/
+WORKDIR /opt/apalache/
+
+# Vendored binaries
+COPY --from=thirdparty /opt/apalache/3rdparty/bin /opt/apalache/3rdparty/bin
+# Vendored libraies
+COPY --from=thirdparty /opt/apalache/3rdparty/lib /opt/apalache/3rdparty/lib
+# The maven repository
+COPY --from=thirdparty /root/.m2 /root/.m2
+
+# Workaround for Surefire not finding ForkedBooter
+# (see https://stackoverflow.com/questions/53010200/maven-surefire-could-not-find-forkedbooter-class)
+ENV _JAVA_OPTIONS="-Djdk.net.URLClassPath.disableClassPathURLCheck=true"
+
+ENV LD_LIBRARY_PATH="/opt/apalache/3rdparty/lib/:${LD_LIBRARY_PATH}"
+
+# clean the leftovers from the previous non-docker builds and build the package
+RUN mvn clean package
+
+FROM openjdk:8-slim
+
+# This is the app image
+#
+# To prepare it, we
+#
+# - copy over all the artifacts needed to run
+# - prepare the enviroment
+# - declare the entrypoint
+
+WORKDIR /opt/apalache/
+
+# Vendored binaries
```

Comment:
Having all these copies isn't pretty, but it's a symptom of us not collecting all our build artifacts under one directory and necessary in the current state, afaict. I'm Not sure if that is idiomatic Java/Scala practice, or just a quirk of our project. But it's something we can address fairly easily in followups if we find it worth while.

Reply:
Ideally, this command should copy the source modules, `bin`, `*.md`, `LICENSE`, and `test`. It should not copy the build-specific and IDE-specific directories such as `target` or `workspace`. This should not be an issue if you publish a docker image from a clean repo.

Reply:
We can have fine-grained copying commands. Java/Scala does not need the compiled classes. Once we have a distribution jar, we need that jar, `src/tla`, `bin`, `3rdparty/{bin,lib}`, and docs, license, etc.

Reply:
>Ideally, this command should copy the source modules, bin, *.md, LICENSE, and test

Could you explain the motive for adding the markdown and test files? iiuc, those shouldn't be needed for the executable in a minimal app image. Do we intend the docker container to be something that users would work inside of, rather than as a way of wrapping the executable?

Did I capture everything you'd advised here?

Reply:
You are right. The docs can be checked at github any time. I don't think the docs and tests are actually required.